packages/ra-core/src/auth/useHandleAuthCallback.ts

```diff
@@ -21,36 +23,62 @@ export const useHandleAuthCallback = (
     const nextPathName = locationState && locationState.nextPathname;
     const nextSearch = locationState && locationState.nextSearch;
     const defaultRedirectUrl = nextPathName ? nextPathName + nextSearch : '/';
+    const { onSuccess, onError, ...queryOptions } = options ?? {};
 
-    return useQuery(
-        ['auth', 'handleCallback'],
-        () => authProvider.handleCallback(),
-        {
-            retry: false,
-            onSuccess: data => {
+    const queryResult = useQuery({
+        queryKey: ['auth', 'handleCallback'],
+        queryFn: () =>
+            authProvider.handleCallback().then(result => result ?? null),
+        retry: false,
+        ...queryOptions,
+    });
+
+    const onSuccessEvent = useEvent(
+        onSuccess ??
+            (() => {
                 // AuthProviders relying on a third party services redirect back to the app can't
                 // use the location state to store the path on which the user was before the login.
                 // So we support a fallback on the localStorage.
                 const previousLocation = localStorage.getItem(
                     PreviousLocationStorageKey
                 );
                 const redirectTo =
-                    (data as AuthRedirectResult)?.redirectTo ??
+                    (queryResult.data as AuthRedirectResult)?.redirectTo ??
                     previousLocation;
-
                 if (redirectTo === false) {
                     return;
                 }
 
                 redirect(redirectTo ?? defaultRedirectUrl);
-            },
-            ...options,
-        }
+            })
     );
+    const onErrorEvent = useEvent(onError ?? noop);
+
+    useEffect(() => {
+        if (queryResult.error == null) return;
+        onErrorEvent(queryResult.error);
+    }, [onErrorEvent, queryResult.error]);
+
+    useEffect(() => {
+        if (queryResult.data === undefined) return;
+        onSuccessEvent(queryResult.data);
```

Comment:
you're passing ythe data to the callback, but the default success callback doesn't use it.